plugins/DevicesDetection/Commands/WarmDeviceDetectorCache.php

```diff
@@ -0,0 +1,148 @@
+
+        $inputFile = $this->openFile(
+            $input->getArgument(self::OPTION_INPUT_FILE), 
+            $input->getOption(self::OPTION_SKIP_HEADER)
+        );
+
+        $maxRowsToProcess = (int)$input->getOption(self::OPTION_ROWS_TO_PROCESS);
+        $counter = 0;
+
+        try {
+            while ($data = fgetcsv($inputFile)) {
+                $counter++;
+                if ($maxRowsToProcess > 0 && $counter > $maxRowsToProcess) {
+                    break;
+                }
+
+                $this->processUserAgent($data[0]);
+            }
+        } finally {
+            fclose($inputFile);
+        }
+        $output->writeln("Written " . $this->numCacheEntriesWritten . " cache entries to file");
+    }
+
+    private function clearCacheDirectory()
+    {
+        $cacheDir = PIWIK_DOCUMENT_ROOT . rtrim(DeviceDetectorCacheEntry::CACHE_DIR, '/');
+        $di = new \RecursiveDirectoryIterator($cacheDir, \FilesystemIterator::SKIP_DOTS);
+        $ri = new \RecursiveIteratorIterator($di, \RecursiveIteratorIterator::CHILD_FIRST);
+        foreach ( $ri as $file ) {
+            $file->isDir() ?  rmdir($file) : unlink($file);
+        }
+    }
+
+    private function openFile($filePath, $skipFirstRow)
+    {
+        if (! file_exists($filePath)) {
+            throw new \Exception("File $filePath not found");
+        }
+        $inputHandle = fopen($filePath, 'r');
+        if ($inputHandle === false) {
+            throw new \Exception("Could not open $filePath");
+        }
+
+        // Skip the first row
+        if ($skipFirstRow) {
+            fgetcsv($inputHandle);
+        }
+        return $inputHandle;
+    }
+
+    private function isValidUserAgentString($userAgent)
+    {
+        $matches = array();
+        foreach (self::$userAgentsPatternsToIgnore as $pattern) {
+            preg_match('/Amazon-Route53-Health-Check-Service[.]*/', $userAgent,   $matches);
+            if ($matches) {
+                return false;
+            }
+        }
+
+        $parts = explode($userAgent, ' ');
+        foreach ($parts as $part) {
+            if (filter_var($part, FILTER_VALIDATE_IP) !== false) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private function processUserAgent($userAgentStr)
+    {
+        $userAgentStr = trim(trim($userAgentStr, '"'));
+        if (!$this->isValidUserAgentString($userAgentStr)) {
+            return;
+        }
+
+        $deviceDetector = DeviceDetectorFactory::getInstance($userAgentStr, false);
+        $outputArray = array(
+            'bot' => $deviceDetector->getBot(),
+            'brand' => $deviceDetector->getBrand(),
+            'client' => $deviceDetector->getClient(),
+            'device' => $deviceDetector->getDevice(),
+            'model' => $deviceDetector->getModel(),
+            'os' => $deviceDetector->getOs()
+        );
+
+        $outputPath = DeviceDetectorCacheEntry::getCachePath($userAgentStr);
```

Comment:
I would suggest to fully exclude detected bots from the caching as they will be excluded from tracking in most cases anyway.

Note: Bot detection is always very fast, as all other detections are skipped if a bot is detected, so there shouldn't be any reason for caching it

Reply:
@sgiehl wouldn't it be still much faster to cache it? In this case it wasn't meant about excluding bots or so but actually to prevent ever caching agents which may contain personal data to prevent them being shipped with Matomo

Reply:
It would be a bit faster. But ihmo there isn't much sense in speeding up tracking requests that won't be tracked at all. And there are other bots that might include "personal" data, like wordpress crawler, which includes the URL

Reply:
> And there are other bots that might include "personal" data, like wordpress crawler, which includes the URL

Awesome, if you know of some that be really helpful so we can prevent shipping them with Matomo ever. Likely won't happen anyway but still good to document 👍